File: TypeScript/app/updateStrategies/AgedBrieStrategy.ts
import Item from "./../models/Item";
import IUpdateStrategy from "./IUpdateStrategy";

export default class AgedBrieStrategy implements IUpdateStrategy {    
    constructor(private qualityFactor : number) {}

    updateItem(item: Item) : Item {
        let quality = Math.min(50, item.quality + 1 * this.qualityFactor);
        let sellIn = item.sellIn - 1;
        return new Item(item.name, sellIn, quality);
    }
}

File: Java/src/main/java/com/gildedrose/strategies/AgedBrieStrategy.java
package com.gildedrose.strategies;

import com.gildedrose.Item;

public enum AgedBrieStrategy implements QualityUpdateStrategy {
    INSTANCE;

    @Override
    public void applyTo(Item item) {
        item.quality = Math.min(50, item.sellIn <= 0 ? item.quality+2 : item.quality+1);
        --item.sellIn;
    }
}
